File: exercise01/BTree.java
public Interface<A, B> Mapper
{
    public B map (A value);
}

public Class<A> BTree
{
    private A value;
    private BTree<A> left;
    private BTree<A> right;

    public BTree(A value, BTree<A> left, BTree<A> right)
    {
        this.value = value;
        this.left = left;
        this.right = right;
    }

    public Tree<B> map(Mapper<A, B> mapper)
    {
        if (left == null && right == null)
            return new Tree<B>(mapper.map(value), null, null);
        if (left == null) 
            return new Tree<B>(mapper.map(value), null, right.map());
        if (right == null)
            return new Tree<B>(mapper.map(value), left.map(), null);
        return new Tree<B>(mapper.map(value), left.map(), right.map());
    }
}

File: exercise01/btree.hs
data BTree a = Node (BTree a) a (BTree a) | Empty

mapTree :: (a -> b) -> BTree a -> BTree b
mapTree f (Node left value right)  = Node (mapTree f left) (f value) (mapTree f right)
mapTree f Empty = Empty

toString :: BTree Int -> String
toString (Node left val right) = "(" ++ toString left ++ " <- " ++ show val ++ " -> " ++ toString right ++ ")"
toString Empty = "_"

display :: BTree Int -> IO ()
display (Node left val right) = print (toString (Node left val right))

main = display (mapTree (\x -> x + 1) (Node (Node Empty 3 Empty) 5 (Node Empty 8 Empty)))